File: android/app/src/main/java/com/targetzone/library/data/repository/AdminRepository.kt
package com.targetzone.library.data.repository

import com.targetzone.library.data.api.ApiClient
import com.targetzone.library.data.model.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody

class AdminRepository {
    private val api = ApiClient.service

    suspend fun getStats(): Result<AdminStats> = runCatching {
        val res = api.getAdminStats()
        res.body()?.data ?: throw Exception("Failed to load stats")
    }

    suspend fun getStudents(page: Int = 0, status: String? = null, membershipStatus: String? = null, search: String? = null, size: Int = 20): Result<Pair<List<StudentSummary>, Int>> = runCatching {
        val res = api.getStudents(page = page, size = size, status = status?.takeIf { it.isNotBlank() }, membershipStatus = membershipStatus?.takeIf { it.isNotBlank() }, search = search?.takeIf { it.isNotBlank() })
        val data = res.body()?.data ?: StudentListResponse()
        Pair(data.students, data.total)
    }

    suspend fun toggleStudentStatus(id: String, active: Boolean): Result<Unit> = runCatching {
        api.toggleStudentStatus(id, ToggleStatusRequest(active = active))
    }

    suspend fun changeSeat(membershipId: String, seatNumber: String): Result<Unit> = runCatching {
        api.changeSeat(membershipId, ChangeSeatRequest(seatNumber = seatNumber))
    }

    suspend fun swapSeat(membershipId: String, otherUserId: String): Result<Unit> = runCatching {
        val res = api.swapSeat(membershipId, SwapSeatRequest(otherUserId = otherUserId))
        // Business-logic failures (e.g. "other student doesn't have a seat") come back
        // as a real 4xx with an { success:false, message } body — Retrofit/Gson only
        // auto-parses res.body() for 2xx, so a non-2xx has to be read from errorBody().
        if (!res.isSuccessful) {
            val message = res.errorBody()?.string()?.let {
                runCatching { com.google.gson.JsonParser.parseString(it).asJsonObject.get("message")?.asString }.getOrNull()
            } ?: "Seat exchange failed"
            throw Exception(message)
        }
        if (res.body()?.success == false) throw Exception(res.body()?.message ?: "Seat exchange failed")
    }

    suspend fun updateMembershipPlan(membershipId: String, planId: String): Result<Unit> = runCatching {
        api.updateMembershipPlan(membershipId, UpdateMembershipPlanRequest(planId = planId))
    }

    suspend fun renewSeat(membershipId: String): Result<Unit> = runCatching {
        api.renewSeat(membershipId)
    }

    suspend fun releaseSeat(membershipId: String, notifyStudent: Boolean): Result<Unit> = runCatching {
        api.releaseSeat(membershipId, ReleaseSeatRequest(notifyStudent))
    }

    suspend fun clearDues(membershipId: String, amountCleared: Double, paymentMode: String = "CASH"): Result<Unit> = runCatching {
        api.clearDues(membershipId, ClearAmountRequest(amountCleared, paymentMode))
    }

    suspend fun markPending(membershipId: String, pendingAmount: Double): Result<Unit> = runCatching {
        api.markPending(membershipId, MarkPendingRequest(pendingAmount))
    }

    suspend fun markGrace(membershipId: String): Result<Unit> = runCatching {
        api.markGrace(membershipId)
    }

    suspend fun deleteStudent(id: String): Result<Unit> = runCatching {
        api.deleteStudent(id)
    }

    suspend fun getStudentsInGraceWithDues(): Result<List<StudentDetail>> = runCatching {
        val res = api.getStudentsInGraceWithDues()
        res.body()?.data ?: emptyList()
    }

    suspend fun getStudentsWithOrphanedSeats(): Result<List<StudentDetail>> = runCatching {
        val res = api.getStudentsWithOrphanedSeats()
        res.body()?.data ?: emptyList()
    }

    suspend fun sendGraceDuesReminders(userIds: List<String>): Result<String> = runCatching {
        val res = api.sendGraceDuesReminders(SendReminderRequest(userIds = userIds))
        res.body()?.data ?: res.body()?.message ?: "Reminders sent"
    }

    suspend fun getExpiringMemberships(withinDays: Int): Result<List<ReminderStudent>> = runCatching {
        val res = api.getExpiringMemberships(withinDays = withinDays)
        res.body()?.data ?: emptyList()
    }

    suspend fun sendReminders(userIds: List<String>): Result<String> = runCatching {
        val res = api.sendReminders(SendReminderRequest(userIds = userIds))
        res.body()?.data ?: res.body()?.message ?: "Reminders sent"
    }

    suspend fun getAdminSeatMap(shift: String, date: String?): Result<List<Seat>> = runCatching {
        val res = api.getAdminSeatMap(shift, date)
        val dto = res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to load seat map")
        dto.seatsByRow.flatMap { (row, seatList) ->
            seatList.map { s ->
                Seat(
                    seatNumber = s.seatNumber,
                    row = row,
                    isBooked = s.isOccupied,
                    studentId = s.studentId,
                    studentName = s.studentName,
                    studentMobile = s.studentMobile,
                    studentGender = s.studentGender,
                    membershipEnd = s.membershipEnd
                )
            }
        }
    }

    suspend fun getStudentDetail(studentId: String): Result<StudentDetail> = runCatching {
        val res = api.getStudentDetail(studentId)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to load student details")
    }

    suspend fun getAllFeedback(): Result<List<FeedbackItem>> = runCatching {
        val res = api.getAllFeedback()
        res.body()?.data ?: emptyList()
    }

    suspend fun updateFeedback(id: String, status: String, adminNotes: String?): Result<FeedbackItem> = runCatching {
        val res = api.updateFeedback(id, UpdateFeedbackRequest(status = status, adminNotes = adminNotes))
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Update failed")
    }

    suspend fun sendBroadcast(message: String, targetGroup: String): Result<String> = runCatching {
        val res = api.sendBroadcast(BroadcastRequest(message = message, targetGroup = targetGroup))
        res.body()?.data ?: res.body()?.message ?: "Broadcast sent"
    }

    suspend fun sendMessageToStudent(id: String, message: String): Result<Unit> = runCatching {
        api.sendMessageToStudent(id, BroadcastRequest(message = message))
    }

    suspend fun createCashMembership(req: CreateCashMembershipRequest): Result<Membership> = runCatching {
        val res = api.createCashMembership(req)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to create membership")
    }

    suspend fun getStudentsWithPendingFees(): Result<List<StudentDetail>> = runCatching {
        val res = api.getStudentsWithPendingFees()
        res.body()?.data ?: emptyList()
    }

    suspend fun clearPendingFees(id: String, amountCleared: Double, paymentMode: String = "CASH"): Result<Unit> = runCatching {
        api.clearPendingFees(id, ClearAmountRequest(amountCleared, paymentMode))
    }

    suspend fun sendPendingFeeReminders(userIds: List<String>): Result<String> = runCatching {
        val res = api.sendPendingFeeReminders(SendReminderRequest(userIds = userIds))
        res.body()?.data ?: res.body()?.message ?: "Reminders sent"
    }

    suspend fun getBroadcastHistory(): Result<List<BroadcastHistory>> = runCatching {
        val res = api.getBroadcastHistory()
        res.body()?.data ?: emptyList()
    }

    suspend fun getRevenueReport(from: String, to: String): Result<RevenueReport> = runCatching {
        val res = api.getRevenueReport(from, to)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to load report")
    }

    suspend fun getDailyPayments(date: String): Result<List<DailyPayment>> = runCatching {
        val res = api.getDailyPayments(date)
        res.body()?.data ?: emptyList()
    }

    suspend fun getInbox(): Result<List<InboxSummary>> = runCatching {
        val res = api.getInbox()
        res.body()?.data ?: emptyList()
    }

    suspend fun getInboxMessage(messageNumber: Int): Result<InboxMessage> = runCatching {
        val res = api.getInboxMessage(messageNumber)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Message not found")
    }

    suspend fun replyToMessage(messageNumber: Int, body: String): Result<Unit> = runCatching {
        api.replyToMessage(messageNumber, ReplyRequest(body))
    }

    suspend fun deleteInboxMessage(messageNumber: Int): Result<Unit> = runCatching {
        api.deleteInboxMessage(messageNumber)
    }

    suspend fun importSingleStudent(name: String, phone: String, photo: java.io.File?): Result<Unit> = runCatching {
        val nameBody = name.toRequestBody("text/plain".toMediaTypeOrNull())
        val phoneBody = phone.toRequestBody("text/plain".toMediaTypeOrNull())
        val photoPart = photo?.let {
            val body = it.asRequestBody("image/jpeg".toMediaTypeOrNull())
            okhttp3.MultipartBody.Part.createFormData("photo", it.name, body)
        }
        val res = api.importSingleStudentWithPhoto(nameBody, phoneBody, photoPart)
        if (res.body()?.success == false) throw Exception(res.body()?.message ?: "Import failed")
    }

    suspend fun updateStudent(id: String, req: UpdateStudentRequest): Result<StudentDetail> = runCatching {
        val res = api.updateStudent(id, req)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Update failed")
    }

    suspend fun importStudents(file: okhttp3.MultipartBody.Part): Result<ImportResult> = runCatching {
        val res = api.importStudents(file)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Import failed")
    }

    suspend fun getExpenses(year: Int, month: Int): Result<MonthlyExpense> = runCatching {
        val res = api.getExpenses(year, month)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to load expenses")
    }

    suspend fun saveExpenses(req: SaveExpenseRequest): Result<MonthlyExpense> = runCatching {
        val res = api.saveExpenses(req)
        res.body()?.data ?: throw Exception(res.body()?.message ?: "Failed to save expenses")
    }

    suspend fun getStudentPayments(userId: String): Result<List<StudentPayment>> = runCatching {
        val res = api.getAdminStudentPayments(userId)
        res.body()?.data ?: emptyList()
    }
}

File: ios/LibrarySystem/LibrarySystem/Features/Admin/AdminRepository.swift
import Foundation

struct AdminRepository {
    static let shared = AdminRepository()
    private let api = APIClient.shared
    private var token: String? { TokenManager.shared.token }

    // MARK: - Dashboard
    func getStats() async throws -> AdminStats {
        try await api.request(.getAdminStats, token: token)
    }

    // MARK: - Students
    func getStudents(page: Int = 0, size: Int = 20, status: String? = nil,
                     membershipStatus: String? = nil, search: String? = nil) async throws -> StudentListResponse {
        try await api.request(.getStudents(page: page, size: size, status: status,
                                           membershipStatus: membershipStatus, search: search),
                              token: token)
    }

    func getStudentDetail(id: String) async throws -> StudentDetail {
        try await api.request(.getStudentDetail(id: id), token: token)
    }

    func getStudentPayments(userId: String) async throws -> [PaymentHistoryItem] {
        try await api.request(.getStudentPayments(userId: userId), token: token)
    }

    func updateStudentPhoto(id: String, photoData: Data) async throws -> String {
        try await api.uploadMultipart(path: "admin/students/\(id)/photo", fieldName: "file",
                                      fileName: "student_\(id).jpg", mimeType: "image/jpeg",
                                      data: photoData, token: token)
    }

    func updateStudent(id: String, req: UpdateStudentRequest) async throws -> StudentDetail {
        try await api.request(.updateStudent(id: id, req: req), token: token)
    }

    func getStudentsWithPendingFees() async throws -> [StudentDetail] {
        try await api.request(.getStudentsWithPendingFees, token: token)
    }

    func getStudentsWithOrphanedSeats() async throws -> [StudentDetail] {
        try await api.request(.getStudentsWithOrphanedSeats, token: token)
    }

    func getStudentSeatHistory(userId: String) async throws -> [SeatHistoryEntry] {
        try await api.request(.getStudentSeatHistory(userId: userId), token: token)
    }

    func deleteStudent(id: String) async throws {
        try await api.requestVoid(.deleteStudent(id: id), token: token)
    }

    func importStudentsCSV(data: Data, fileName: String, mimeType: String) async throws -> ImportResult {
        try await api.uploadMultipartDecoded(path: "admin/students/import", fieldName: "file",
                                             fileName: fileName, mimeType: mimeType,
                                             data: data, token: token)
    }

    func importSingleStudentWithPhoto(name: String, phone: String, photoData: Data) async throws -> ManualImportWithPhotoResponse {
        try await api.uploadMultipartWithFields(
            path: "admin/students/import/single-with-photo",
            fields: ["name": name, "phone": phone],
            fileFieldName: "photo", fileName: "student.jpg", mimeType: "image/jpeg",
            fileData: photoData, token: token)
    }

    func clearPendingFees(userId: String, amountCleared: Double, paymentMode: String? = nil) async throws {
        let req = ClearAmountRequest(amountCleared: amountCleared, paymentMode: paymentMode)
        try await api.requestVoid(.clearPendingFees(userId: userId, req: req), token: token)
    }

    func getStudentsInGraceWithDues() async throws -> [StudentDetail] {
        try await api.request(.getStudentsInGraceWithDues, token: token)
    }

    func clearDues(membershipId: String, amountCleared: Double, paymentMode: String? = nil) async throws {
        let req = ClearAmountRequest(amountCleared: amountCleared, paymentMode: paymentMode)
        try await api.requestVoid(.clearDues(membershipId: membershipId, req: req), token: token)
    }

    func releaseSeat(membershipId: String, notifyStudent: Bool) async throws {
        let req = ReleaseSeatRequest(notifyStudent: notifyStudent)
        try await api.requestVoid(.releaseSeat(membershipId: membershipId, req: req), token: token)
    }

    func markPending(membershipId: String, pendingAmount: Double) async throws {
        let req = MarkPendingRequest(pendingAmount: pendingAmount)
        try await api.requestVoid(.markPending(membershipId: membershipId, req: req), token: token)
    }

    func markGrace(membershipId: String) async throws {
        try await api.requestVoid(.markGrace(membershipId: membershipId), token: token)
    }

    func renewSeat(membershipId: String) async throws {
        try await api.requestVoid(.renewSeat(membershipId: membershipId), token: token)
    }

    func runExpiryCheck() async throws -> String {
        try await api.request(.runExpiryCheck, token: token)
    }

    func getSeatHistory(seatNumber: String) async throws -> [SeatHistoryEntry] {
        try await api.request(.getSeatHistory(seatNumber: seatNumber), token: token)
    }

    // MARK: - Memberships & Seats
    func changeSeat(membershipId: String, seatNumber: String) async throws {
        let req = ChangeSeatRequest(seatNumber: seatNumber)
        try await api.requestVoid(.changeSeat(membershipId: membershipId, req: req), token: token)
    }

    func updateMembershipPlan(membershipId: String, planId: String) async throws {
        let req = UpdateMembershipPlanRequest(planId: planId)
        try await api.requestVoid(.updateMembershipPlan(membershipId: membershipId, req: req), token: token)
    }

    func getExpiringMemberships(withinDays: Int = 7) async throws -> [ReminderStudent] {
        try await api.request(.getExpiringMemberships(withinDays: withinDays), token: token)
    }

    // admin-service's MembershipDto for this endpoint has no planType field (and
    // none of the other callers of this method use the returned membership either
    // — the response body is only ever a success/failure signal in practice), so
    // decoding it against the full Membership model (which requires planType)
    // failed every time: the membership was created successfully server-side, but
    // the client-side decode of the response threw, surfacing as a "Data error"
    // instead of the actual success message. requestVoid sidesteps the mismatch
    // entirely rather than making Membership.planType optional for every other
    // caller that genuinely relies on it being present.
    func createCashMembership(studentId: String, planId: String, seatNumber: String,
                              shift: String, startDate: String,
                              paidAmount: Double?, pendingAmount: Double?,
                              paymentMode: String? = nil) async throws {
        let req = CreateCashMembershipRequest(studentId: studentId, planId: planId,
                                              seatNumber: seatNumber, shift: shift,
                                              startDate: startDate, paidAmount: paidAmount,
                                              pendingAmount: pendingAmount, paymentMode: paymentMode)
        try await api.requestVoid(.createCashMembership(req), token: token)
    }

    func getAdminSeatMap(shift: String, date: String? = nil) async throws -> SeatMapDto {
        try await api.request(.getAdminSeatMap(shift: shift, date: date), token: token)
    }

    // MARK: - Reminders & Broadcast
    func sendReminders(userIds: [String]) async throws {
        let req = SendReminderRequest(userIds: userIds)
        try await api.requestVoid(.sendReminders(req), token: token)
    }

    func sendPendingFeeReminders(userIds: [String]) async throws {
        let req = SendReminderRequest(userIds: userIds)
        try await api.requestVoid(.sendPendingFeeReminders(req), token: token)
    }

    func sendGraceDuesReminders(userIds: [String]) async throws {
        let req = SendReminderRequest(userIds: userIds)
        try await api.requestVoid(.sendGraceDuesReminders(req), token: token)
    }

    // MARK: - Notification Settings
    func getNotificationSettings() async throws -> [NotificationSetting] {
        try await api.request(.getNotificationSettings, token: token)
    }

    func updateNotificationSetting(key: String, sendToStudent: Bool, sendToAdmin: Bool,
                                   hindiEnabled: Bool, hindiTextStudent: String?,
                                   hindiTextAdmin: String?) async throws -> NotificationSetting {
        let req = UpdateNotificationSettingRequest(sendToStudent: sendToStudent, sendToAdmin: sendToAdmin,
                                                   hindiEnabled: hindiEnabled,
                                                   hindiTextStudent: hindiTextStudent,
                                                   hindiTextAdmin: hindiTextAdmin)
        return try await api.request(.updateNotificationSetting(key: key, req: req), token: token)
    }

    func sendBroadcast(message: String, targetGroup: String = "ALL") async throws {
        let req = BroadcastRequest(message: message, targetGroup: targetGroup)
        try await api.requestVoid(.sendBroadcast(req), token: token)
    }

    func sendMessageToStudent(id: String, message: String) async throws {
        let req = BroadcastRequest(message: message, targetGroup: "ALL")
        try await api.requestVoid(.sendMessageToStudent(id, req), token: token)
    }

    func getBroadcastHistory() async throws -> [BroadcastHistory] {
        try await api.request(.getBroadcastHistory, token: token)
    }

    // MARK: - Feedback
    func getAllFeedback() async throws -> [FeedbackItem] {
        try await api.request(.getAllFeedback, token: token)
    }

    func updateFeedback(id: String, status: String, adminNotes: String?) async throws -> FeedbackItem {
        let req = UpdateFeedbackRequest(status: status, adminNotes: adminNotes)
        return try await api.request(.updateFeedback(id: id, req: req), token: token)
    }

    // MARK: - Revenue Reports
    func getRevenueReport(from: String, to: String) async throws -> RevenueReport {
        try await api.request(.getRevenueReport(from: from, to: to), token: token)
    }

    func getDailyPayments(date: String) async throws -> [DailyPayment] {
        try await api.request(.getDailyPayments(date: date), token: token)
    }

    func getPaymentBreakdown(from: String, to: String) async throws -> [PaymentBreakdownItem] {
        try await api.request(.getPaymentBreakdown(from: from, to: to), token: token)
    }

    // MARK: - App Settings
    func getAppSettings() async throws -> AppSettings {
        try await api.request(.getAppSettings, token: token)
    }

    func saveAppSettings(wifiName: String?, wifiPassword: String?, upiId: String?, graceDays: Int,
                        convenienceFee: Double, waterTankerRate: Double, couponsEnabled: Bool) async throws -> AppSettings {
        let req = SaveAppSettingsRequest(wifiName: wifiName, wifiPassword: wifiPassword, upiId: upiId,
                                         graceDays: graceDays, convenienceFee: convenienceFee,
                                         waterTankerRate: waterTankerRate, couponsEnabled: couponsEnabled)
        return try await api.request(.saveAppSettings(req), token: token)
    }

    // MARK: - Coupons (admin CRUD)
    func getCoupons() async throws -> [Coupon] {
        try await api.request(.getAdminCoupons, token: token)
    }

    func createCoupon(code: String?, discountPercent: Int) async throws -> Coupon {
        let req = CreateCouponRequest(code: code, discountPercent: discountPercent)
        return try await api.request(.createCoupon(req), token: token)
    }

    func updateCoupon(id: String, discountPercent: Int? = nil, isActive: Bool? = nil) async throws -> Coupon {
        let req = UpdateCouponRequest(discountPercent: discountPercent, isActive: isActive)
        return try await api.request(.updateCoupon(id: id, req: req), token: token)
    }

    func deleteCoupon(id: String) async throws {
        try await api.requestVoid(.deleteCoupon(id: id), token: token)
    }

    // MARK: - UPI Pay Links & Payment Verification
    func createPayLink(studentId: String, amount: Double) async throws -> String {
        let req = PayLinkRequest(studentId: studentId, amount: amount)
        let res: PayLinkResponse = try await api.request(.createPayLink(req), token: token)
        return res.link
    }

    func getPaymentClaims(status: String? = nil) async throws -> [PaymentClaim] {
        try await api.request(.getPaymentClaims(status: status), token: token)
    }

    func reviewPaymentClaim(id: String, status: String) async throws -> PaymentClaim {
        let req = PaymentClaimReviewRequest(status: status)
        return try await api.request(.reviewPaymentClaim(id: id, req: req), token: token)
    }

    // MARK: - Expenses
    func getExpenses(year: Int? = nil, month: Int? = nil) async throws -> MonthlyExpense {
        try await api.request(.getExpenses(year: year, month: month), token: token)
    }

    func saveExpenses(year: Int, month: Int, waterTankerQty: Int, waterTankerPrice: Double,
                      electricityBill: Double, internetBill: Double,
                      miscItems: [MiscItemRequest]) async throws -> MonthlyExpense {
        let req = SaveExpenseRequest(year: year, month: month, waterTankerQty: waterTankerQty,
                                     waterTankerPrice: waterTankerPrice, electricityBill: electricityBill,
                                     internetBill: internetBill, miscItems: miscItems)
        return try await api.request(.saveExpenses(req), token: token)
    }

    // MARK: - Inbox
    func getInbox() async throws -> [InboxSummary] {
        try await api.request(.getInbox, token: token)
    }

    func getInboxMessage(_ number: Int) async throws -> InboxMessage {
        try await api.request(.getInboxMessage(number), token: token)
    }

    func replyToMessage(_ number: Int, body: String) async throws {
        let req = ReplyRequest(body: body)
        try await api.requestVoid(.replyToMessage(number, req: req), token: token)
    }

    func deleteInboxMessage(_ number: Int) async throws {
        try await api.requestVoid(.deleteInboxMessage(number), token: token)
    }

    // MARK: - Plans
    func getPlans() async throws -> [Plan] {
        try await api.request(.getPlans)
    }
}
